imports/lib/models/Model.ts

```diff
@@ -0,0 +1,465 @@
+    } = {},
+  ): Promise<number> {
+    const { bypassSchema = false, ...mongoOptions } = options;
+
+    // Note that Meteor's update implementation will drop options that it
+    // doesn't recognize (including bypassDocumentValidation), so we need to go
+    // around it and do some of its work ourselves.
+    //
+    // However, Meteor tries pretty hard to make sure its ID generation scheme
+    // behaves correctly in the presence of upserts, and that's too much work to
+    // do ourselves, so ban upserts with bypassSchema. The behavior can be
+    // replicated easily enough by trying to insert, and falling back on an
+    // update if that fails.
+    if (bypassSchema) {
+      if (options.upsert) {
+        throw new Error('Cannot bypass schema validation when upserting');
+      }
+
+      try {
+        const result = await this.collection.rawCollection().updateOne(
+          typeof selector === 'object' ? selector : { _id: selector },
+          modifier,
+          {
+            ...mongoOptions,
+            bypassDocumentValidation: true,
+          } as any
+        );
+        return result.modifiedCount;
+      } catch (e) {
+        formatValidationError(e);
+        throw e;
+      }
+    }
+
+    const parsed = await parseMongoModifierAsync(this.relaxedSchema, modifier, !!options.upsert);
+    try {
+      return await this.collection.updateAsync(selector, parsed, mongoOptions);
+    } catch (e) {
+      formatValidationError(e);
+      throw e;
+    }
+  }
+
+  // See the comments around bypassSchema in updateAsync for why we don't
+  // support bypassSchema here
+  async upsertAsync(
+    selector: Selector<z.output<this['schema']>>,
+    modifier: Exclude<Mongo.Modifier<z.output<this['schema']>>, z.output<this['schema']>>,
+    options: {
+      multi?: boolean | undefined;
+    } = {},
+  ): Promise<{
+    numberAffected?: number | undefined;
+    insertedId?: z.output<typeof stringId> | undefined;
+  }> {
+    const parsed = await parseMongoModifierAsync(this.relaxedSchema, modifier, true);
+    try {
+      return await this.collection.upsertAsync(selector, parsed, options) as any;
+    } catch (e) {
+      formatValidationError(e);
+      throw e;
+    }
+  }
+
+  async removeAsync(
+    selector: Selector<z.output<this['schema']>>,
+  ): Promise<number> {
+    return this.collection.removeAsync(selector);
+  }
+
+  // For now, don't allow transforms on query methods. There's no fundamental
+  // reason not to, but I wasn't able to get the types to work
+
+  find<
+    S extends Selector<z.output<this['schema']>>,
+    O extends Omit<Mongo.Options<z.output<this['schema']>>, 'transform'>,
+  >(
+    selector?: S,
+    options?: O,
+  ) {
+    return this.collection.find(selector, options) as
+      Mongo.Cursor<SelectorToResultType<z.output<this['schema']>, S>>;
+  }
+
+  findOne<S extends Selector<z.output<this['schema']>>, O extends Omit<Mongo.Options<z.output<this['schema']>>, 'limit' | 'transform'>>(
```

Comment:
I guess we still need all the sync methods for now because we use them in client-side code?

I wonder what Meteor's plans are re: deprecating/removing the sync methods from the minimongo implementation or just leaving them in.  If they eliminate them, then we have a bunch of frontend code rewriting to do, and I'm not sure that will play well with React where we generally expect synchronous results.

Reply:
Yeah, that's right. I haven't seen any discussion about what the expected client-side behavior is, but my assumption is that minimongo will continue to be synchronous (there's no real reason for it not to be)